### PR TITLE
CORE-5458 Flow Error Handling

### PR DESCRIPTION
**Pull request by simon-johnson-r3**
- Made FlowFiberImpl more robust
- Made FlowEventPipelineImpl time out based on configuration so at least the FlowWorker can become active again
-- Defaults to 1 minute
- Wrote and then disabled FlowFiberImpl test, see CORE-4820

### CODE REVIEW

File: applications/tools/flow-worker-setup/config.conf [hunk 1]
```diff
@@ -19,11 +19,6 @@ corda {
             heartbeatTimeout = 100000
             p2pTTL = 60000
         },
-        processing {
```
Comment (simon-johnson-r3) [inline applications/tools/flow-worker-setup/config.conf:22]: Unrelated to this change, but I spotted some incorrect duplication as I was adding the new config property

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FiberFuture.kt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package net.corda.flow.fiber
+
+import java.util.concurrent.Future
+
+class FiberFuture(
```
Comment (simon-johnson-r3) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FiberFuture.kt:5]: you can't interrupt futures, there's no link from them to whatever it is they're executing, so we need our own class to encapsulate all this functionality as a single returned object from running a Flow

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/factory/FlowFiberFactoryImpl.kt [hunk 3]
```diff
@@ -24,10 +22,15 @@ class FlowFiberFactoryImpl : FlowFiberFactory {
         ScheduledSingleThreadExecutor()
     )
 
-    override fun createFlowFiber(flowId: String, logic: FlowLogicAndArgs): FlowFiber {
```
Comment (simon-johnson-r3) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/factory/FlowFiberFactoryImpl.kt:27]: This touches on the tech debt ticket https://r3-cev.atlassian.net/browse/CORE-5721
Starting and resuming a Flow have unnecessarily different apis, coincidentally by adding a requirement to expose interruptions it starts to make them more closely aligned.

File: components/flow/flow-service/src/test/kotlin/net/corda/flow/fiber/FlowFiberImplTest.kt [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+import org.mockito.kotlin.clearInvocations
+import org.mockito.kotlin.mock
+import org.mockito.kotlin.verify
+import org.mockito.kotlin.whenever
+import java.util.*
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.Future
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+@Disabled
```
Comment (simon-johnson-r3) [inline components/flow/flow-service/src/test/kotlin/net/corda/flow/fiber/FlowFiberImplTest.kt]: Here's the disabled test which would pass if you apply the Quasar agent to the unit tests, https://r3-cev.atlassian.net/browse/CORE-4820

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/impl/FlowEventPipelineImpl.kt [hunk 5]
```diff
@@ -138,16 +140,27 @@ class FlowEventPipelineImpl(
...
             outcome
         )
 
-        /*
-        Need to think about a timeout for the get(), what do we do if a flow does not complete?
-        */
-        when (val flowResult = flowResultFuture.get()) {
+        val flowResult = try {
+            flowResultFuture.future.get(timeoutMilliseconds, TimeUnit.MILLISECONDS)
+        } catch (e: TimeoutException) {
```
Comment (owenstanford) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/impl/FlowEventPipelineImpl.kt:154]: I couldn't see a test that covers the timeout path

File: components/flow/flow-service/src/test/kotlin/net/corda/flow/fiber/FlowFiberImplTest.kt [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        flowFiber.setUncaughtExceptionHandler { _, _ ->
+            // This is never hit if the test passes, but the test would hang without it were the test to fail
+            forceCancelOngoingRunUntilWithTestFailure()
+        }
+        val outcome = flowFiber.runUntilNotSuspended(mockFlowFiberExecutionContext, fiberScheduler)
+
+        assertThat(outcome).isInstanceOfAny(FlowIORequest.FlowFailed::class.java)
+        verify(mockFlowLogic).invoke()
+    }
+
```
Comment (owenstanford) [inline components/flow/flow-service/src/test/kotlin/net/corda/flow/fiber/FlowFiberImplTest.kt:81]: Is It possible to test the suspend outcomes as part of this test?

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/impl/FlowEventPipelineImpl.kt (hunk 5 above)
Comment (simon-johnson-r3) [reply components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/impl/FlowEventPipelineImpl.kt:154]: now done, https://github.com/corda/corda-runtime-os/pull/1707/files#diff-a8de5ac133260d07e65b12ca537db3165bba6714aee2bb5cf027d24ff2cdf5f2 line 195 onwards

**Review by owenstanford — APPROVED** [review]

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FlowFiberImpl.kt [hunk 7]
```diff
@@ -51,28 +51,41 @@ class FlowFiberImpl(
...
+            runFlow()
+        } catch (t: Throwable) {
+            log.error("FlowFiber failed due to internal Throwable being thrown", t)
+            failTopLevelSubFlow(t)
+        }
+
+        if (!flowCompletion.isDone) {
+            log.error("runFlow failed to complete normally, forcing a failure")
+            failTopLevelSubFlow(IllegalStateException("Flow failed to complete normally, forcing a failure"))
+        }
```
Comment (lankydan) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FlowFiberImpl.kt:68]: Is this change required, or is it just some refactoring?

If it's refactoring, personally I actually find it harder to follow myself.

If its a required change, then I'm not seeing why at the moment (my bad).

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/impl/FlowEventPipelineImpl.kt [hunk 8]
```diff
@@ -138,16 +140,27 @@ class FlowEventPipelineImpl(
...
+            log.error("Flow execution timeout, Flow marked as failed, interrupt attempted")
+            // This works in extremely limited circumstances. The Flow which experienced a timeout will continue to
+            // show the status RUNNING. Flows started in the waiting period will end up stuck in the START_REQUESTED
+            // state. The biggest benefit to this timeout is the error logging and the fact that waiting here
+            // indefinitely is blocking the FlowWorker for all Flows indefinitely too. See CORE-5820.
+            flowResultFuture.interruptable.attemptInterrupt()
```
Comment (lankydan) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/impl/FlowEventPipelineImpl.kt:160]: Can you quickly talk me through this one.

**Review by lankydan — COMMENTED** [review]
Two comments, the refactoring one is my opinion, if you disagree its fine. Want to talk about the other one since I think I'm missing a bit of understanding.

Otherwise fine.

**Review by lankydan — APPROVED** [review]
